iocore/net/quic/QUICFrameRetransmitter.h

```diff
@@ -0,0 +1,120 @@
+/** @file
+ *
+ *  A brief file description
+ *
+ *  @section license License
+ *
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <deque>
+#include "QUICFrame.h"
+
+class QUICFrameGenerator;
+
+struct QUICFrameInformation {
+  QUICFrameType type;
+  QUICEncryptionLevel level;
+
+  uint8_t data[128] = {};
+};
+
+struct QUICFrameInformationDeleter {
+  void operator()(QUICFrameInformation *p);
+};
+
+using QUICFrameInformationUPtr = std::unique_ptr<QUICFrameInformation, QUICFrameInformationDeleter>;
+
+constexpr QUICFrameType RETRANSMITTED_FRAME_TYPE[] = {QUICFrameType::STREAM, QUICFrameType::STOP_SENDING,
+                                                      QUICFrameType::RST_STREAM};
+
+struct StreamFrameInfo {
+  QUICStreamId stream_id;
+  QUICOffset offset;
+  bool has_fin;
+  Ptr<IOBufferBlock> block;
+};
+
+struct MaxStreamDataFrameInfo {
+  QUICStreamId stream_id;
+  uint64_t maximum_stream_data;
+};
+
+struct RstStreamFrameInfo {
+  QUICStreamId stream_id;
+  QUICAppErrorCode error_code;
+  QUICOffset final_offset;
+};
+
+struct StopSendingFrameInfo {
+  QUICStreamId stream_id;
+  QUICAppErrorCode error_code;
+};
+
+struct MaxDataFrameInfo {
+  uint64_t maximum_data;
+};
+
+struct BlockedFrameInfo {
+  QUICOffset offset;
+};
+
+struct StreamBlockedFrameInfo {
+  QUICStreamId stream_id;
+  QUICOffset offset;
+};
+
+struct CryptoFrameInfo {
+  QUICOffset offset;
+  Ptr<IOBufferBlock> block;
+};
+
+struct AckFrameInfo {
+  QUICPacketNumber largest_acknowledged;
+};
+
+class QUICFrameRetransmitter
+{
+public:
+  virtual QUICFrameUPtr create_retransmitted_frame(QUICEncryptionLevel level, uint16_t maximum_frame_size, QUICFrameId id = 0,
+                                                   QUICFrameGenerator *owner = nullptr);
+  virtual void save_frame_info(QUICFrameInformationUPtr &info);
+
+private:
+  QUICFrameUPtr _create_stream_frame(QUICFrameInformationUPtr &info, uint16_t maximum_frame_size,
+                                     std::deque<QUICFrameInformationUPtr> &tmp_queue, QUICFrameId id, QUICFrameGenerator *owner);
+  QUICFrameUPtr _create_reset_stream_frame(QUICFrameInformationUPtr &info, uint16_t maximum_frame_size,
```

Comment:
I wrote only about MAX_DATA frames explicitly in my previous comment but I think RST_STREAM and STOP_SENDING don't need to use this retransmitter too. If you looked into the detail, you could find `_is_reset_sent` and `_is_stop_sending_sent `. QUICStream has enough information to regenerate the frames.

Reply:
Actually I only wants to introduce this class . So there is no code inherited from it.  The frames retransmittion  will be completed in #4770 .

Reply:
And I'm sorry to make that not so clear.